### ISSUE TITLE
after localeSubpaths: true, SSR can't find route

### ISSUE THREAD

**Issue by andylacko** [isaachinman/next-i18next#88]
after switching to lang with non naked domain eg. /sk/xxx, after refresh I get 404 not found

example piece from my server.js (configuration is exactly copied from example server.js file):

```javascript
server.get('/offers', (req, res) => app.render(req, res, '/offers', req.params))

server.get('*', (req, res) => handle(req, res))
```

PS: don't know, if this lib wraps somehow links that are not defined by handle()

**Comment by isaachinman** [comment]
Yes, that's correct. We currently don't have a good way of supporting custom routing like this. What happens if you put the `nextI18NextMiddleware` line _after_ your custom route declaration?

**Comment by capellini** [comment]
@andylacko using [this issue comment as a guide](https://github.com/isaachinman/next-i18next/issues/62#issuecomment-449875049), you could do something like this (full disclosure, I haven't even ran the below code to see if it works, but you can understand the substance of what I'm trying to do):

```js
(async () => {
  const app = next({ dev: process.env.NODE_ENV !== 'production' })

  const handleCustomI18nRoutes = (req, res, pathname, query) => {
    if (pathname === '/offers') {
      app.render(req, res, '/offers', { ...query, ...req.params })
    } else {
      app.render(req, res, pathname, query)
    }
  }

  await app.prepare()
  const server = express()

  nextI18NextMiddleware(i18next, { render: handleCustomI18nRoutes }, server)

  server.get('/offers', (req, res) => app.render(req, res, '/offers', req.params))
  server.get('*', (req, res) => handler(req, res))

  await server.listen(3000)
  console.log('> Ready on http://localhost:3000')
})()
```
Basically, the only thing that `nextI18NextMiddleware` does with app (at the moment) is call `app.render` [here](https://github.com/isaachinman/next-i18next/blob/e2f507865d50397fc38e375e90a00ed009d2a1eb/src/middlewares/next-i18next-middleware.js#L21) if `localeSubpaths` is true and the path matches a non-default language path (i.e. is prefixed with the locale).  All we're doing is intercepting that call and adding our custom routing with the actual `app.render` call, if needed.

It's not ideal, as there is some code duplication (you still need to `server.get('/offers', ...)` to listen for GET requests in the default language), but I believe that something along these lines should work.

@isaachinman given the large amount of interest, would you mind if I created a few more sites under the `examples/` directory that included some custom routing?  My idea is that I could do two -- one that uses `next-routes` and one which has a custom route handler.

I realize that we're still developing the API, so things may change, but we could always evolve the examples as they do.  Also, I could put a disclaimer in the `README` that states we don't officially support custom routes at this time and that this is a workaround.  But I think that it could cut down on the amount of issues submitted.

**Comment by isaachinman** [comment]
@capellini If I understand your suggestion correctly, that would not work with `localeSubpaths` at all.

Instead of creating more examples, I'd like it if we can come up with a systematic and simple way of supporting custom routing. The simplest thing I can think of is to simply expose the `lngPathDetector` and `lngPathCorrector` utils, but I don't think that's best.

Instead, it'd be great to expose some sort of `addCustomRoute` function that takes the route as the first argument and a `req, res, next` function as the second argument, or something similar.

Any ideas?

**Comment by capellini** [comment]
> that would not work with `localeSubpaths` at all

I just added a `custom-routing-example` branch to my fork where I created a very trivial example.  Apologies for writing code that I didn't even try -- it is a bit more involved than the code that I wrote above and I may be missing something, as the example is quite trivial, but I'm currently doing something similar for my custom routing.  But it is very kludgy and there has to be a better way.

> Any ideas?

I don't know all of the internal workings of `i18next`, `react-i18next`, and `next-i18next` very intimately, but what if instead of calling `app.render` in `nextI18NextMiddleware`, it acts as middleware and modifies the appropriate fields in the request object?  It seems that all [this line](https://github.com/isaachinman/next-i18next/blob/e2f507865d50397fc38e375e90a00ed009d2a1eb/src/middlewares/next-i18next-middleware.js#L21) is doing is adding a `lng` query parameter and removing the locale from the url.  So what if we did something like the following instead?

```js
    server.get(`/:lng(${allLanguages.join('|')})/*`, (req, res, done) => {
      const { lng } = req.params

      req.query = { ...req.query, lng }
      req.url = url.replace(`/${lng}`, '')
      done()
    })
```
Again, the above example is probably not 100% correct, but this way, the user would call `nextI18NextMiddleware` before any other `server.use/get/...` calls in their server code and the req object would be modified accordingly without any other modifications needed.

**Comment by isaachinman** [comment]
@capellini That does seem sensible at first glance, and would side-step a lot of problems. Do you have time to put together an initial approach? I am quite busy with other things at the moment. If not, I will get to this perhaps next week. Thanks!
